### PR TITLE
Implementation of Naive Tridiagonal Linear Solver (Thomas' Algorithm) in C++

### LINKED ISSUE #88: Implement Tridiagonal Linear System Solver in C++
Translate the fortran code for solving a linear system consisting of a tridiagonal matrix to C++. Add tests to verify the accuracy of the solution. 

**Acceptance Criteria**
1. Add large values to the diagonal elements of a matrix so that the conditional number is low.
2. Make sure the relative error is smaller than 1.e-7 for single precision and 1.e-15 for double precision.

### PR DESCRIPTION
**Pull request by AdityaDendukuri**
**Implementation Notes**
1. Implemented the tridiagonal matrix algorithm. 
2. This is a Naive implementation, so no pivoting to ensure stability.

**Testing Notes**
1. Testing done by uniformly sampling random matrix (A) and vector (x), and checking the reconstruction error (L2 norm) between x and tridiag_solve(A, A*x). 
2. Matrix generated is diagonally dominant as there is no partial pivoting for ensuring stability. 
4. Testing done for both double and single precision data types. 

Next Steps - Compare speed, stability and accuracy of the implementation with LAPACK's subroutine for tridiagonal solver. 

closes #88

### CODE REVIEW

File: test/unit/linalg/tridiag.cpp [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+  std::ofstream file;
+  file.open("../../../tool/numerics/errors_dbl.txt");
+  for (int i = 0; i < n_trials; i++) {
+    file << errors[i] << std::endl;
+  }
+  file.close();
+
+  EXPECT_EQ(failed, 0);
+}
+*/
```
Comment (K20shores) [inline test/unit/linalg/tridiag.cpp]: This seems like a useful test. However, we don't like to accept large blocks of commented code. Please either remove this test, uncomment the code block and add a description of what it's trying to test and turn it into a real test, or copy and paste this code into an issue describing what would need to be done to fix this test and delete the test.

File: include/tuvx/linalg/linalg.h [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+template <typename T> void fill_rand_vec(std::vector<T> &x, int size);
+
+template <typename T> void fill_rand_mat(trid_mat<T> &trid_mat, int n);
+
+template <typename T> void print_vec(std::vector<T>);
+
+template <typename T> void print_trid_mat(trid_mat<T> x);
+
+template <typename T>
+std::vector<T> tridiag_solve(trid_mat<T> A, std::vector<T> b);
```
Comment (K20shores) [inline include/tuvx/linalg/linalg.h]: Matt setup several other types `Array2D`, `Array3D` that we will be using to hold the bulk of the data for operational use in TUV (unless you make a stronger case for eigen or some other matrix library). Is this code expected to work with that type or would we do a slight refactor when we go to use this code?

File: include/tuvx/linalg/linalg.inl [hunk 3]
```diff
@@ -0,0 +1,83 @@
+
+#include <vector>
```
Comment (K20shores) [inline include/tuvx/linalg/linalg.inl]:
```suggestion
```
Since this file is included directly in the `linalg.h` header, this isn't necessary here. However, you're free to leave it

File: include/tuvx/linalg/linalg.h [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+namespace linalg {
+
+// tridiag matrix
+template <typename T> struct trid_mat {
+  int size;
+  std::vector<T> udiag; // upper diagonal
+  std::vector<T> ldiag; // lower diagonal
+  std::vector<T> mdiag; // main diagonal
+};
+
```
Comment (K20shores) [inline include/tuvx/linalg/linalg.h]: Please add some doxygen style comments to these functions. If you're using vscode with the c++ plugins, if you type three slashes `///` it will auto create the base doxygen comments and all you have to do is fill out the details

File: include/tuvx/linalg/linalg.inl [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+namespace tuvx {
+namespace linalg {
+
+template <typename T>
+inline std::vector<T> dot(trid_mat<T> A, std::vector<T> x) {
+  int size = x.size();
```
Comment (K20shores) [inline include/tuvx/linalg/linalg.inl]:
```suggestion
  std::size_t size = x.size();
```
or use `auto` for the type. In either case, this reduces compiler warnings about losing precision when we use these sizes in other places

File: include/tuvx/linalg/linalg.inl [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+  std::vector<T> v(size);
+  v[0] = A.mdiag[0] * x[0] + A.udiag[0] * x[1];
+
+  int i = 0;
+  for (i = 1; i < x.size() - 1; i++) {
+    v[i] =
+        A.mdiag[i] * x[i] + A.udiag[i] * x[i + 1] + A.ldiag[i - 1] * x[i - 1];
+  }
+  v[i] = A.mdiag[i] * x[i] + A.ldiag[i - 1] * x[i - 1];
```
Comment (K20shores) [inline include/tuvx/linalg/linalg.inl]:
```suggestion
  std::size_t i = 0;
  for (i = 1; i < size - 1; i++) {
    v[i] =
        A.mdiag[i] * x[i] + A.udiag[i] * x[i + 1] + A.ldiag[i - 1] * x[i - 1];
  }
  v[i] = A.mdiag[i] * x[i] + A.ldiag[i - 1] * x[i - 1];
```

Since you defined `size` above, may as well use it here

File: include/tuvx/linalg/linalg.inl [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+  return v;
+}
+
+template <typename T>
+inline std::vector<T> tridiag_solve(trid_mat<T> A, std::vector<T> b) {
+  T temp;
+  int N = b.size();
```
Comment (K20shores) [inline include/tuvx/linalg/linalg.inl]:
```suggestion
  std::size_t N = b.size();
```

File: include/tuvx/linalg/linalg.inl [hunk 8]
```diff
@@ -0,0 +1,83 @@
...
+  std::vector<T> x(N);
+  // forward pass
+  for (int i = 1; i < N; i++) {
```
Comment (K20shores) [inline include/tuvx/linalg/linalg.inl]:
```suggestion
  for (std::size_t i = 1; i < N; i++) {
```

File: include/tuvx/linalg/linalg.h [hunk 9]
```diff
@@ -0,0 +1,35 @@
...
+#include <iostream>
+#include <random>
+#include <vector>
+
+namespace tuvx {
+namespace linalg {
+
+// tridiag matrix
+template <typename T> struct trid_mat {
+  int size;
```
Comment (K20shores) [inline include/tuvx/linalg/linalg.h]:
```suggestion
  std::size_t size;
```
It may help to be more consistent with what the stl does, and this will prevent negative values for the size

**Review by K20shores — CHANGES_REQUESTED** [review]
Overall it looks good! I left a few comments related to using `std::size_t` instead of ints. I prefer to use this for anything that indexes into an stl container since the `.size` method of all stl containers returns a `std::size_t`. I didn't suggest that you change it in every case, only the cases that are on the hot path for actual scientific work. It may help us reduce compiler warnings and prevent typecasts from happening automatically.

File: include/tuvx/linalg/linalg.h (hunk 2 above)
Comment (AdityaDendukuri) [reply include/tuvx/linalg/linalg.h]: Since the tridiagonal matrix is sparse, I only store the diagonal values in three std::vectors  instead of the full matrix in `Array2D`, as the rest of the entries are zero. The function is tailored to work with this specific sparse representation as the algorithm is only designed for tridiagonal forms. Although I could refactor it to accept three `Array1D` arguments instead (`linpak%tridiag` is designed like this too).

File: include/tuvx/cross_section.hpp
Comment (sjsprecious) [inline include/tuvx/cross_section.hpp:1]: @K20shores @mattldawson this is maybe a question for you. I don't believe Adi made this change and it should come from the `main` branch. Why does it show as a change here?

File: include/tuvx/cross_section_params.hpp
Comment (sjsprecious) [inline include/tuvx/cross_section_params.hpp:1]: the same question above.

File: include/tuvx/linalg/linalg.h (hunk 4 above)
Comment (sjsprecious) [reply include/tuvx/linalg/linalg.h]: @AdityaDendukuri Also do `udiag`, `ldiag` and `mdiag` have the same dimension size? Theoretically `mdiag` should have one more element than the other two. We probably want to double check this and make sure that the same input can be used in LAPACK and DPLASMA directly later.

File: test/unit/linalg/tridiag.cpp
Comment (sjsprecious) [inline test/unit/linalg/tridiag.cpp:1]: Not sure if this is the naming convention but we may want a test file beginning with the prefix `test_`.

File: include/tuvx/cross_section.hpp
Comment (K20shores) [reply include/tuvx/cross_section.hpp:1]: I think this happens when you branch off of a an old branch that gets merged into main only *after* you branched. I don't understand it, or how to fix it.

File: include/tuvx/linalg/linalg.h (hunk 4 above)
Comment (AdityaDendukuri) [reply include/tuvx/linalg/linalg.h]: @sjsprecious No they do not. If the matrix is $n \times n$, `udiag` and `ldiag` will be `n-1` sized vectors. I considered making them same size by padding with `0`, which might enhance the readability of the code. I will check how LAPACKE handles this.

File: test/unit/linalg/tridiag.cpp
Comment (AdityaDendukuri) [reply test/unit/linalg/tridiag.cpp:1]: I changed this from `test_tridiag.cpp` because cmake renames the test target as `test_test_tridiag`. I assumed cmake adds the `test_` prefix.

File: include/tuvx/linalg/linalg.h (hunk 4 above)
Comment (sjsprecious) [reply include/tuvx/linalg/linalg.h]: Awesome. Hopefully we can use the same interface for those standard math libraries and the rest part of TUV-x.

File: test/unit/linalg/tridiag.cpp
Comment (sjsprecious) [reply test/unit/linalg/tridiag.cpp:1]: @K20shores is this something special in TUV-x? In MICM all our test files starting with `test_` though.

File: include/tuvx/cross_section.hpp
Comment (sjsprecious) [reply include/tuvx/cross_section.hpp:1]: Oh, good to know it and thanks for the information.

File: include/tuvx/linalg/linalg.h [hunk 10]
```diff
@@ -0,0 +1,35 @@
+#pragma once
+
+#include <iostream>
+#include <random>
+#include <vector>
+
+namespace tuvx {
+namespace linalg {
```
Comment (mattldawson) [inline include/tuvx/linalg/linalg.h]: I don't think we've settled on a strategy for when to introduce a new namespace (aside from having one for the library as a whole). I'm fine with introducing one for linear algebra, but am curious about what the rest of the group thinks and whether we should formalize some approach to introducing namespaces?

File: include/tuvx/linalg/linalg.h (hunk 2 above)
Comment (K20shores) [reply include/tuvx/linalg/linalg.h]: Ok, `std::vector` seems fine then. We can always refactor this later if we decide we need something fancier

File: include/tuvx/linalg/linalg.h [hunk 11]
```diff
@@ -0,0 +1,35 @@
...
+
+// tridiag matrix
+template <typename T> struct trid_mat {
```
Comment (mattldawson) [inline include/tuvx/linalg/linalg.h]:
```suggestion
template <typename T> struct TridiagonalMatrix {
```
We've been trying to follow the [Google C++ Style Guide](https://google.github.io/styleguide/cppguide.html#General_Naming_Rules) and using whole words and specific styles for `ClassNames`, `class_members_`, `FunctionsNames()`, `function_arguments`

File: include/tuvx/linalg/linalg.h [hunk 12]
```diff
@@ -0,0 +1,35 @@
...
+  int size;
+  std::vector<T> udiag; // upper diagonal
+  std::vector<T> ldiag; // lower diagonal
+  std::vector<T> mdiag; // main diagonal
```
Comment (mattldawson) [inline include/tuvx/linalg/linalg.h]:
```suggestion
  std::vector<T> upper_diagonal_; // upper diagonal
  std::vector<T> lower_diagonal_; // lower diagonal
  std::vector<T> main_diagonal_; // main diagonal
```

File: include/tuvx/linalg/linalg.h [hunk 13]
```diff
@@ -0,0 +1,35 @@
...
+};
+
+template <typename T> void fill_rand_vec(std::vector<T> &x, int size);
```
Comment (mattldawson) [inline include/tuvx/linalg/linalg.h]:
```suggestion
/// @brief Allocates a vector to a given size and fills it with random values
/// @param x Vector to allocate and fill
/// @param size Size of the vector
template <typename T> void FillRandomVector(std::vector<T> &x, int size);
```
We've been using Doxygen to document the code for MICM. @K20shores @sjsprecious - should we do the same for TUV-x C++ code?
(and similar for other functions here...)

**Review by mattldawson — CHANGES_REQUESTED** [review]
Very nice! And great work on setting up the tests!

My only suggestions are on naming and style. I added comments on some specific lines, but you'll want to look over the rest of the code for similar places to update.

File: include/tuvx/linalg/linalg.h (hunk 13 above)
Comment (sjsprecious) [reply include/tuvx/linalg/linalg.h]: I like this idea and we probably should do it consistently for TUV-x as well. It is a good time to mention this since we just begin our work.

File: test/unit/linalg/CMakeLists.txt [hunk 14]
```diff
@@ -0,0 +1,11 @@
+################################################################################
+# Test utilities
+
+include(test_util)
+
+################################################################################
+# tests
+
+create_standard_cxx_test(NAME test_tridiag SOURCES tridiag.cpp)
```
Comment (sjsprecious) [inline test/unit/linalg/CMakeLists.txt]: Based on the discussion with @K20shores , you should change the name from `test_tridiag` to `tridiag` so that you do not get `test_test_` shown up in the testing later.

File: include/tuvx/linalg/linalg.h (hunk 11 above)
Comment (AdityaDendukuri) [reply include/tuvx/linalg/linalg.h]: Got it. I will make those changes.

File: test/unit/linalg/tridiag.cpp
Comment (sjsprecious) [reply test/unit/linalg/tridiag.cpp:1]: @AdityaDendukuri after discussing with Kyle, you can name this file beginning with `test_` and should update the `test/unit/linalg/CMakeLists.txt` file to avoid the issue of duplicated `test_` prefix.

File: test/unit/linalg/tridiag.cpp (hunk 1 above)
Comment (AdityaDendukuri) [reply test/unit/linalg/tridiag.cpp]: I have decided to remove it. There are always some edge cases where the precision errors cause the test to fail (for example an error of 1.1e-7 not satisfying the error tolerance of 1e-7). I did reduce the tolerance to 1e-5 as suggested by Dr. Cao.

File: include/tuvx/linalg/linalg.h (hunk 13 above)
Comment (AdityaDendukuri) [reply include/tuvx/linalg/linalg.h]: @mattldawson @sjsprecious @K20shores I have added Doxygen style documentation to my code. Let me know if I did it correctly.

File: include/tuvx/linalg/linalg.h (hunk 4 above)
Comment (AdityaDendukuri) [reply include/tuvx/linalg/linalg.h]: @sjsprecious I have checked the LAPACK implementation. They do use $n-1$ sized vectors for ldiag and udiag. However, they pass in these three diagonals separately into the function rather than a combined structure like I do. Let me know if you think I should write it that way too.

File: include/tuvx/linalg/linalg.h (hunk 10 above)
Comment (K20shores) [reply include/tuvx/linalg/linalg.h]: I don't think it's necessary; since we will be using full names `tridiag_solve`, rather than like 3 characters (looking at you lapack), we won't be conflicting with any packages we bring in, probably

Comment (AdityaDendukuri) [reply include/tuvx/linalg/linalg.h]: I have removed the linalg namespace specification

File: include/tuvx/linalg/linalg.h (hunk 4 above)
Comment (sjsprecious) [reply include/tuvx/linalg/linalg.h]: Thanks for checking it. As long as the dimension is the same, we can change the interface later as needed. No need to worry about it right now.

File: include/tuvx/linalg/linalg.h (hunk 13 above)
Comment (sjsprecious) [reply include/tuvx/linalg/linalg.h]: i think you did it correctly and I appreciate those comments.

**Comment by AdityaDendukuri** [comment]
Hi everyone, thank you for the feedback! I think I addressed all of your comments. Let me know if I missed anything.

**Comment by sjsprecious** [comment]
> Hi everyone, that you for the feedback! I think I addressed all of your comments. Let me know if I missed anything.

Hi @AdityaDendukuri , thanks for addressing those comments. If you look through Kyle's comments above, there are still some unaddressed comments related to the `int` type used in your current code. Can you change them to be `std::size_t` or `auto` as @K20shores suggested?

**Review by K20shores — APPROVED** [review]

File: cmake/dependencies.cmake [hunk 15]
```diff
@@ -1,32 +1,33 @@
 find_package(PkgConfig REQUIRED)
 include(FetchContent)
 
-################################################################################
+# ##############################################################################
```
Comment (mattldawson) [inline cmake/dependencies.cmake:4]: It looks like somehow a lot of formatting changes ended up in this file. Could you checkout this file from the main branch?

File: include/tuvx/linalg/linalg.h [hunk 16]
```diff
@@ -0,0 +1,62 @@
...
+  std::size_t size;
+  std::vector<T> upper_diagonal_; // upper diagonal
+  std::vector<T> lower_diagonal_; // lower diagonal
+  std::vector<T> main_diagonal_;  // main diagonal
+};
+
+/// @brief Allocates a vector to a given size and fills it with random values
+/// @param x Vector to allocate and fill
+/// @param size Size of the vector
+template <typename T> void FillRandVec(std::vector<T> &x, std::size_t size);
```
Comment (mattldawson) [inline include/tuvx/linalg/linalg.h]:
```suggestion
template <typename T> void FillRandomVector(std::vector<T> &x, std::size_t size);
```
formatting looks good - just want to use full words in names

File: include/tuvx/linalg/linalg.h [hunk 17]
```diff
@@ -0,0 +1,62 @@
...
+
+/// @brief Allocates a matrix to a given size and fills it with random values
+/// @param A tridiagonal matrix to allocate and fill
+/// @param size Size of the vector
+template <typename T> void FillRandMat(TridiagonalMatrix<T> &A, std::size_t n);
```
Comment (mattldawson) [inline include/tuvx/linalg/linalg.h]:
```suggestion
template <typename T> void FillRandomMatrix(TridiagonalMatrix<T> &A, std::size_t n);
```

File: include/tuvx/linalg/linalg.h [hunk 18]
```diff
@@ -0,0 +1,62 @@
...
+
+/// @brief displays the data stored inside a std::vector
+/// @param x Vector to print
+template <typename T> void PrintVec(std::vector<T>);
```
Comment (mattldawson) [inline include/tuvx/linalg/linalg.h]:
```suggestion
template <typename T> void PrintVector(std::vector<T>);
```

File: include/tuvx/linalg/linalg.h [hunk 19]
```diff
@@ -0,0 +1,62 @@
...
+
+/// @brief displays the data stored inside a tridiagonal matrix.
+/// For now, this function simply calls printvec() on the three diagonals.
+/// @param A tridiagonal matrix to print
+template <typename T> void PrintTridMat(TridiagonalMatrix<T> x);
```
Comment (mattldawson) [inline include/tuvx/linalg/linalg.h]:
```suggestion
template <typename T> void PrintTridiagonalMatrix(TridiagonalMatrix<T> x);
```

File: include/tuvx/linalg/linalg.h [hunk 20]
```diff
@@ -0,0 +1,62 @@
...
+
+/// @fn Tridiagonal Linear System Solver
+/// @brief Thomas' algorithm for solving tridiagonal linear system (A x = b)
+/// @param A tridiagonal coeffcient matrix
+/// @param b right hand side vector of the tridiagonal system.
+/// @returns x solution that solves A x = b.
+template <typename T>
+std::vector<T> TridiagSolve(TridiagonalMatrix<T> A, std::vector<T> b);
```
Comment (mattldawson) [inline include/tuvx/linalg/linalg.h]:
```suggestion
std::vector<T> TridiagonalSolve(TridiagonalMatrix<T> A, std::vector<T> b);
```

**Review by mattldawson — CHANGES_REQUESTED** [review]
Looking better! Just a couple last style things

File: cmake/dependencies.cmake (hunk 15 above)
Comment (AdityaDendukuri) [reply cmake/dependencies.cmake:4]: This was because I added the Google Benchmark library here (although I'm not certain if this is the correct method for adding a dependency). I removed it because benchmarking and comparing my code to LAPACK will be addressed in a new issue and pull request.

File: include/tuvx/linalg/linalg.h (hunk 16 above)
Comment (AdityaDendukuri) [reply include/tuvx/linalg/linalg.h]: Done! Let me know if I missed anything.

File: include/tuvx/linalg/linalg.inl [hunk 21]
```diff
@@ -0,0 +1,85 @@
...
+    v[i] = A.main_diagonal_[i] * x[i] + A.upper_diagonal_[i] * x[i + 1] +
+           A.lower_diagonal_[i - 1] * x[i - 1];
+  }
+  v[i] = A.main_diagonal_[i] * x[i] + A.lower_diagonal_[i - 1] * x[i - 1];
+  return v;
+}
+
+template <typename T>
+inline std::vector<T> TridiagonalSolve(TridiagonalMatrix<T> A,
+                                       std::vector<T> b) {
```
Comment (boulderdaze) [inline include/tuvx/linalg/linalg.inl]: Could you make the parameters that doesn't need a copy to `const ref` ? This request applies to all the functions in this PR. 
```suggestion
template <typename T>
inline std::vector<T> TridiagonalSolve(const TridiagonalMatrix<T>& A,
                                       const std::vector<T>& b) {
```

File: include/tuvx/cross_section.hpp [hunk 22]
```diff
@@ -0,0 +1,7 @@
+
+class tuvx_cross_section {
+  private:
+    
+
+
+};
```
Comment (boulderdaze) [inline include/tuvx/cross_section.hpp:7]: Is this working in progress? Would it make sense that you submit this when it's complete?

File: include/tuvx/cross_section_params.hpp [hunk 23]
```diff
@@ -0,0 +1,10 @@
+#pragma once 
+
+#include <string>
+
+#include <vector>
+
+namespace tuvx {
+  public:
+     
+}
```
Comment (boulderdaze) [inline include/tuvx/cross_section_params.hpp:10]: Is this working in progress? Would it make sense that you submit this when it's complete?

File: include/tuvx/linalg/linalg.inl [hunk 24]
```diff
@@ -0,0 +1,85 @@
...
+  T temp;
+  std::size_t N = b.size();
+  std::vector<T> x(N);
+  // forward pass
+  for (std::size_t i = 1; i < N; i++) {
+    temp = A.lower_diagonal_[i - 1] / A.main_diagonal_[i - 1];
+    A.main_diagonal_[i] -= temp * A.upper_diagonal_[i - 1];
+    b[i] -= temp * b[i - 1];
+  }
+  // back substitution
```
Comment (boulderdaze) [inline include/tuvx/linalg/linalg.inl]: Can you fix the typo?

File: include/tuvx/linalg/linalg.inl [hunk 25]
```diff
@@ -0,0 +1,85 @@
...
+    nrm += std::pow(x[i], order);
+  }
+  return (1.0 / nrm) * std::pow(nrm, 1.0 / order);
+}
+
+template <typename T> inline void PrintVector(std::vector<T> x) {
+  std::cout << std::endl;
+  for (int i = 0; i < (int)x.size(); i++) {
+    std::cout << x.at(i) << std::endl;
+  }
```
Comment (boulderdaze) [inline include/tuvx/linalg/linalg.inl]: Would it be possible to make the style consistent in `for loop` across the code base? I don't think you need  cast here
```suggestion
  for (int i = 0; i < x.size(); i++) {
    std::cout << x.at(i) << std::endl;
  }
```

File: include/tuvx/linalg/linalg.inl [hunk 26]
```diff
@@ -0,0 +1,85 @@
...
+}
+
+template <typename T> inline void Norm(std::vector<T> x, int order) {
+  T nrm = (T)0;
+  for (auto i = 0; i < x.size(); i++) {
+    nrm += std::pow(x[i], order);
+  }
+  return (1.0 / nrm) * std::pow(nrm, 1.0 / order);
+}
+
```
Comment (boulderdaze) [inline include/tuvx/linalg/linalg.inl]: Could you add test for this function?

**Review by boulderdaze — CHANGES_REQUESTED** [review]
Looks good, just have a few change requests, hopefully easy to fix.

**Review by mattldawson — APPROVED** [review]
looks great! I don't see anything other than what others have commented on

File: include/tuvx/cross_section.hpp (hunk 22 above)
Comment (AdityaDendukuri) [reply include/tuvx/cross_section.hpp:7]: I did not edit this file. Not sure how it got added to my PR

File: include/tuvx/linalg/linalg.inl (hunk 21 above)
Comment (AdityaDendukuri) [reply include/tuvx/linalg/linalg.inl]: The `TridiagonalSolve` code modifies the arguments instead of defining new temporary variables in the function. The rest of the functions in this PR do not do this so I will apply your suggestions to them.

**Review by sjsprecious — APPROVED** [review]
Thanks @AdityaDendukuri for working on this PR. It looks great to me.

All the 74 tests pass on Derecho using gcc/12 compiler.

**Review by mattldawson — APPROVED** [review]
looks great!

File: include/tuvx/linear_algebra/linear_algebra.inl [hunk 27]
```diff
@@ -0,0 +1,99 @@
...
+template <typename T>
+T ComputeError(const std::vector<T> &x, const std::vector<T> &x_approx,
+               int norm_order) {
+  T error = 0;
+  for (std::size_t i = 0; i < x.size(); i++) {
+    error += std::pow(std::abs(x[i] - x_approx[i]), norm_order);
+  }
+  error = ((T)1 / x.size()) * std::pow(error, (T)1 / (T)norm_order);
+  return error;
+}
```
Comment (mwaxmonsky) [inline include/tuvx/linear_algebra/linear_algebra.inl]: Probably not a concern now but we might want to refactor this in the future to handle `T=int` or other non-floating point types.

File: include/tuvx/linear_algebra/linear_algebra.inl [hunk 28]
```diff
@@ -0,0 +1,99 @@
...
+  }
+  // back substitution
+  x[N - 1] = b[N - 1] / A.main_diagonal_[N - 1];
+  for (std::size_t i = N - 2;; i--) {
+    x[i] = (b[i] - A.upper_diagonal_[i] * x[i + 1]) / A.main_diagonal_[i];
+    if (i == 0) {
+      break;
+    }
+  }
+  return x;
```
Comment (mwaxmonsky) [inline include/tuvx/linear_algebra/linear_algebra.inl:47]: Does this seem a little more readable?
```suggestion
  for (std::size_t i = N - 2; i >= 0; i--) {
    x[i] = (b[i] - A.upper_diagonal_[i] * x[i + 1]) / A.main_diagonal_[i];
  }
  return x;
```

File: test/unit/linear_algebra/test_error_function.cpp [hunk 29]
```diff
@@ -0,0 +1,62 @@
...
+/// testing the Linear approximation solvers
+TEST(ErrorFunctionTest, ErrorSinglePrecision) {
+
+  // same vector should return 0 error
+  vecf x(size);
+  vecf x1(size);
+  FillRandom<float>(x);
+  x1 = x;
+  float error = ComputeError<float>(x, x1, norm_order);
+  EXPECT_EQ(error, (float)0);
```
Comment (mwaxmonsky) [inline test/unit/linear_algebra/test_error_function.cpp]: Is there a style preference we have for `(float)0` vs `0.0f`?  `0.0{f,d}` would also match the naming convention for the vector/matrix types as well.

File: include/tuvx/linear_algebra/linear_algebra.inl [hunk 30]
```diff
@@ -0,0 +1,99 @@
...
+  }
+  error = ((T)1 / x.size()) * std::pow(error, (T)1 / (T)norm_order);
+  return error;
+}
+
+template <typename T> inline void Print(const TridiagonalMatrix<T> &x) {
+  std::cout << "----" << std::endl;
+  print_vec(x.upper_diagonal_);
+  print_vec(x.main_diagonal_);
+  print_vec(x.lower_diagonal_);
```
Comment (mwaxmonsky) [inline include/tuvx/linear_algebra/linear_algebra.inl]: Might cause a compile error if called?
```suggestion
  Print(x.upper_diagonal_);
  Print(x.main_diagonal_);
  Print(x.lower_diagonal_);
```

**Review by mwaxmonsky — CHANGES_REQUESTED** [review]
Great work @AdityaDendukuri!  Just one compile issue and a few things we might want to take a second look at but nothing major!

File: include/tuvx/linear_algebra/linear_algebra.inl (hunk 28 above)
Comment (AdityaDendukuri) [reply include/tuvx/linear_algebra/linear_algebra.inl:47]: @mwaxmonsky I originally implemented this loop like you suggested, but it was causing segmentation faults as the std::size_t index was being reset to a large value. I had a discussion with @mattldawson and @sjsprecious regarding this and we went with this fix.

Comment (mwaxmonsky) [reply include/tuvx/linear_algebra/linear_algebra.inl:47]: Yup, my apologies, just talked with everyone and they explained the behavior to me so this is perfect!

File: include/tuvx/linear_algebra/linear_algebra.inl [hunk 31]
```diff
@@ -0,0 +1,99 @@
...
+}
+
+template <typename T>
+T ComputeError(const std::vector<T> &x, const std::vector<T> &x_approx,
+               int norm_order) {
+  T error = 0;
+  for (std::size_t i = 0; i < x.size(); i++) {
+    error += std::pow(std::abs(x[i] - x_approx[i]), norm_order);
+  }
+  error = ((T)1 / x.size()) * std::pow(error, (T)1 / (T)norm_order);
```
Comment (sjsprecious) [inline include/tuvx/linear_algebra/linear_algebra.inl]: Should this be `error = std::pow(error / x.size(), (T)1 / (T)norm_order)`? I am thinking RMSE may be a better measurement than the L2 norm here. What do you think @mwaxmonsky ?

Also we can compute the relative error (i.e., `sum(abs(x[i] - x_approx[i])/max(x[i], x_approx[i]))/x.size()`) and compare it to the machine epsilon.

Comment (mwaxmonsky) [reply include/tuvx/linear_algebra/linear_algebra.inl]: @sjsprecious I agree with using RMSE here.  Unless there is an integration issue with the rest of the tuv-x code that depends on the LP norm, I think RMSE makes more sense in this context.

**Review by K20shores — APPROVED** [review]

**Review by boulderdaze — APPROVED** [review]

File: include/tuvx/linear_algebra/linear_algebra.inl (hunk 31 above)
Comment (AdityaDendukuri) [reply include/tuvx/linear_algebra/linear_algebra.inl]: @sjsprecious I am not familiar with the relative error code you shared. Isn't the relative error defined as
$$\frac{||x - x'||_p}{||x||_p}$$
where $x'$ is an approximation for $x$?

Comment (AdityaDendukuri) [reply include/tuvx/linear_algebra/linear_algebra.inl]: @sjsprecious @mwaxmonsky I have used the relative error instead of absolute in my latest commit. I have also used the error tolerance to be the machine epsilon instead of setting it arbitrarily. Lastly, I added a test checking the consistency in error over various system sizes.

File: test/unit/linear_algebra/test_tridiagonal_solver.cpp [hunk 32]
```diff
@@ -16,10 +16,10 @@ typedef std::vector<double> vecd;
...
 typedef std::vector<float> vecf;
 
-const std::size_t size = 100000000; // size of the system to test
-const double tol_dp = 1e-14;        // tolorance for double
-const float tol_sp = 1e-5; // tolorance for single point floating precision
-const int norm_order = 1;  // L2 norm for computing error
+const std::size_t size = 10; // size of the system to test
+const double tol_dp = std::numeric_limits<double>::epsilon();
+const float tol_sp = std::numeric_limits<float>::epsilon();
+const int norm_order = 2; // L2 norm for computing error
```
Comment (sjsprecious) [inline test/unit/linear_algebra/test_tridiagonal_solver.cpp]: if you change to / stay with the relative error, probably the `norm_order` option is no longer needed?

File: test/unit/linear_algebra/test_error_function.cpp [hunk 33]
```diff
@@ -0,0 +1,62 @@
...
+
+  // L1 norm between [0.1 0 0 0 ...] and [0 0 0 0 0] should be 0.1/size;
+  std::fill(x.begin(), x.end(), (float)0);
+  std::fill(x1.begin(), x1.end(), (float)0);
```
Comment (sjsprecious) [inline test/unit/linear_algebra/test_error_function.cpp]: Since we are testing the relative error, initializing the values with 0 will lead to a NAN result. Need to update the tests here.

**Review by sjsprecious — CHANGES_REQUESTED** [review]

File: test/unit/linear_algebra/test_error_function.cpp (hunk 33 above)
Comment (AdityaDendukuri) [reply test/unit/linear_algebra/test_error_function.cpp]: @sjsprecious  Fixed in the latest commit!

File: test/unit/linear_algebra/test_error_function.cpp [hunk 34]
```diff
@@ -0,0 +1,64 @@
+
+#include <cfloat>
+#include <cstddef>
+#include <cstdlib>
+#include <gtest/gtest.h>
+#include <tuvx/linear_algebra/linear_algebra.hpp>
+#include <vector>
+
+const std::size_t size = 10; // size of the system to test
+const int norm_order = 2;    // L2 norm for computing error
```
Comment (sjsprecious) [inline test/unit/linear_algebra/test_error_function.cpp]: If we switch to use the relative error, probably we do not need the `norm_order` variable in your test any more. If so, please clean up all the code and comments about it.

**Review by sjsprecious — CHANGES_REQUESTED** [review]

File: test/unit/linear_algebra/test_error_function.cpp (hunk 34 above)
Comment (AdityaDendukuri) [reply test/unit/linear_algebra/test_error_function.cpp]: @sjsprecious I have removed everything in reference to `norm_order`.

File: test/unit/linear_algebra/test_error_function.cpp (hunk 29 above)
Comment (AdityaDendukuri) [reply test/unit/linear_algebra/test_error_function.cpp]: @mwaxmonsky I have replaced all `(double)0` with `0.0`, `(float)0` with `0.0f` in my latest commit. This seems cleaner.

Comment (mwaxmonsky) [reply test/unit/linear_algebra/test_error_function.cpp]: Thanks @AdityaDendukuri!

**Review by mwaxmonsky — APPROVED** [review]
Looks great @AdityaDendukuri!

**Review by sjsprecious — APPROVED** [review]
Thanks @AdityaDendukuri for addressing all the comments. Great work!
